src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -0,0 +1,1926 @@
+
+		newTargetEntry->expr = (Expr *) copyObject(column);
+		appendStringInfo(columnNameString, WORKER_COLUMN_FORMAT, resNo);
+		newTargetEntry->resname = columnNameString->data;
+		newTargetEntry->resjunk = false;
+		newTargetEntry->resno = resNo;
+
+		subqueryTargetEntryList = lappend(subqueryTargetEntryList, newTargetEntry);
+		resNo++;
+	}
+
+	return subqueryTargetEntryList;
+}
+
+
+/*
+ * UpdateVarMappingsForExtendedOpNode updates varno/varattno fields of columns
+ * in columnList to point to corresponding target in subquery target entry
+ * list.
+ */
+static void
+UpdateVarMappingsForExtendedOpNode(List *columnList, List *subqueryTargetEntryList)
+{
+	ListCell *columnCell = NULL;
+	foreach(columnCell, columnList)
+	{
+		Var *columnOnTheExtendedNode = (Var *) lfirst(columnCell);
+		ListCell *targetEntryCell = NULL;
+		foreach(targetEntryCell, subqueryTargetEntryList)
+		{
+			TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+			Var *targetColumn = NULL;
+
+			Assert(IsA(targetEntry->expr, Var));
+			targetColumn = (Var *) targetEntry->expr;
+			if (columnOnTheExtendedNode->varno == targetColumn->varno &&
+				columnOnTheExtendedNode->varattno == targetColumn->varattno)
+			{
+				columnOnTheExtendedNode->varno = 1;
+				columnOnTheExtendedNode->varattno = targetEntry->resno;
+				break;
+			}
+		}
+	}
+}
+
+
+/*
+ * MultiSubqueryPushdownTable creates a MultiTable from the given subquery,
+ * populates column list and returns the multitable.
+ */
+static MultiTable *
+MultiSubqueryPushdownTable(Query *subquery)
+{
+	MultiTable *subqueryTableNode = NULL;
+	StringInfo rteName = makeStringInfo();
+	List *columnNamesList = NIL;
+	ListCell *targetEntryCell = NULL;
+
+	appendStringInfo(rteName, "worker_subquery");
+
+	foreach(targetEntryCell, subquery->targetList)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		columnNamesList = lappend(columnNamesList, makeString(targetEntry->resname));
+	}
+
+	subqueryTableNode = CitusMakeNode(MultiTable);
+	subqueryTableNode->subquery = subquery;
+	subqueryTableNode->relationId = SUBQUERY_PUSHDOWN_RELATION_ID;
+	subqueryTableNode->rangeTableId = SUBQUERY_RANGE_TABLE_ID;
+	subqueryTableNode->partitionColumn = NULL;
+	subqueryTableNode->alias = makeNode(Alias);
+	subqueryTableNode->alias->aliasname = rteName->data;
+	subqueryTableNode->referenceNames = makeNode(Alias);
+	subqueryTableNode->referenceNames->aliasname = rteName->data;
+	subqueryTableNode->referenceNames->colnames = columnNamesList;
+
+	return subqueryTableNode;
+}
+
+
+/*
+ * QueryPushdownSqlTaskList creates a list of SQL tasks to execute the given subquery
+ * pushdown job. For this, the it is being checked whether the query is router
+ * plannable per target shard interval. For those router plannable worker
+ * queries, we create a SQL task and append the task to the task list that is going
+ * to be executed.
+ */
+List *
+QueryPushdownSqlTaskList(Query *query, uint64 jobId,
```

Comment:
Here we're mixing physical planner logic with logical planning logic. It'd be nice to preserve the separation. Maybe we should introduce another file?

Reply:
I have moved `QueryPushdownSqlTaskList` and related functions to multi_physical_planner (as you said as an option in our meeting yesterday), otherwise I need to introduce a new file with three functions only. After https://github.com/citusdata/citus/pull/2124, this logic will be used by router_planner to create tasks for modify queries with multiple range table entries.